projects/rocr-runtime/rocrtst/suites/functional/counted_queues.cc

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright © Advanced Micro Devices, Inc., or its affiliates.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "suites/functional/counted_queues.h"
+#include "hsa/hsa_ext_amd.h"
+#include "hsa/hsa.h"
+#include "common/base_rocr_utils.h"
+#include "gtest/gtest.h"
+
+
+CountedQueuesTest::CountedQueuesTest() : TestBase() {
+  set_num_iteration(10);  // Number of iterations to execute of the main test;
+                          // This is a default value which can be overridden
+                          // on the command line.
+
+  set_title("RocR Counted Queues Test");
+  set_description(
+      "This test validates the behavior of Shared Counted Queues managed by the "
+      "Counted Queue Manager in a scenario where both HIP and OpenMP utilize CP "
+      "Queues, causing oversubscription and a subsequent performance degradation.");
+}
+
+CountedQueuesTest::~CountedQueuesTest() {}
+
+void CountedQueuesTest::SetUp() { TestBase::SetUp(); }
+
+void CountedQueuesTest::Run() {
+  // Compare required profile for this test case with what we're actually
+  // running on
+  if (!rocrtst::CheckProfile(this)) {
+    return;
+  }
+  TestBase::Run();
+}
+
+void CountedQueuesTest::Close() {
+  // This will close handles opened within rocrtst utility calls and call
+  // hsa_shut_down(), so it should be done after other hsa cleanup
+  TestBase::Close();
+}
+
+void CountedQueuesTest::DisplayResults() const {
+  // Compare required profile for this test case with what we're actually
+  // running on
+  if (!rocrtst::CheckProfile(this)) {
+    return;
+  }
+  TestBase::DisplayResults();
+}
+
+void CountedQueuesTest::DisplayTestInfo() { TestBase::DisplayTestInfo(); }
+
+void CountedQueuesTest::CountedQueueBasicApiTest() {
+  hsa_status_t status;
+
+  // Find all gpu agents
+  std::vector<hsa_agent_t> gpus;
+  status = hsa_iterate_agents(rocrtst::IterateGPUAgents, &gpus);
+  ASSERT_EQ(status, HSA_STATUS_SUCCESS);
+
+  hsa_queue_t* queue = nullptr;
+  status = hsa_amd_counted_queue_acquire(
+      gpus[0], HSA_QUEUE_TYPE_MULTI, HSA_AMD_QUEUE_PRIORITY_NORMAL, nullptr, nullptr, 0, &queue);
+
+  ASSERT_EQ(status, HSA_STATUS_SUCCESS);
+  ASSERT_NE(queue, nullptr);
+
+  // Query counted queue and check internal reference count
+  uint32_t use_count = 0;
+  status = hsa_amd_counted_queue_get_info(queue, HSA_QUEUE_INFO_USE_COUNT, &use_count);
+  ASSERT_EQ(status, HSA_STATUS_SUCCESS);
+  EXPECT_EQ(use_count, 1);  // should be 4 after acquire
+
+  // Release the queue
+  status = hsa_amd_counted_queue_release(queue);
+  ASSERT_EQ(status, HSA_STATUS_SUCCESS);
```

Comment:
We can query HSA_QUEUE_INFO_USE_COUNT after hsa_amd_counted_queue_release(...) and confirm that the use count is back to 0.